src/kubelet/client/connector.go

```diff
@@ -0,0 +1,252 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/transport"
+
+	"github.com/newrelic/nri-kubernetes/v2/src/client"
+)
+
+const (
+	apiProxyPath = "/api/v1/nodes/%s/proxy/"
+	httpSchema   = "http"
+	httpsSchema  = "https"
+)
+
+type Connector interface {
+	connect() (*connParams, error)
+}
+
+type defaultConnector struct {
+	// TODO: Use a non-sdk logger
+	logger             log.Logger
+	apiServerHost      string
+	kc                 kubernetes.Interface
+	tripperBearerToken http.RoundTripper
+	config             config.Mock
+}
+
+func (dp defaultConnector) connect() (*connParams, error) {
+
+	kubeletPort, err := dp.getKubeletPort()
+	if err != nil {
+		return nil, fmt.Errorf("getting kubelet port: %w", err)
+	}
+
+	kubeletSchema := dp.getKubeletSchema(kubeletPort)
+	hostURL := net.JoinHostPort(dp.config.NodeIP, fmt.Sprint(kubeletPort))
+
+	conn, err := dp.setupLocalConnection(dp.tripperBearerToken, kubeletSchema, hostURL)
+	if err == nil {
+		dp.logger.Debugf("connected to Kubelet directly with nodeIP")
+		return conn, nil
+	}
+
+	dp.logger.Infof("Kubelet connection with nodeIP not working, falling back to API proxy: %v", err)
+
+	conn, err = dp.checkConnectionAPIProxy(dp.apiServerHost, dp.config.NodeName, dp.tripperBearerToken)
+	if err != nil {
+		return nil, fmt.Errorf("creating connection parameters for API proxy: %w", err)
+	}
+
+	return conn, nil
+}
+
+func (dp defaultConnector) setupLocalConnection(tripperWithBearerToken http.RoundTripper, schema string, hostURL string) (*connParams, error) {
+	dp.logger.Debugf("connecting to kubelet directly with nodeIP")
+	var err error
+	var conn *connParams
+
+	switch schema {
+	case httpSchema:
+		if conn, err = dp.checkConnectionHTTP(hostURL); err == nil {
+			return conn, nil
+		}
+	case httpsSchema:
+		if conn, err = dp.checkConnectionHTTPS(hostURL, tripperWithBearerToken); err == nil {
+			return conn, nil
+		}
+	default:
+		// we were not able to infer the schema and the user did not provided it.
+		if conn, err = dp.checkConnectionHTTPS(hostURL, tripperWithBearerToken); err == nil {
+			return conn, nil
+		}
+
+		if conn, err = dp.checkConnectionHTTP(hostURL); err == nil {
+			return conn, nil
+		}
+	}
+
+	return nil, fmt.Errorf("no connection succeeded through localhost: %w", err)
+}
+
+func (dp defaultConnector) getKubeletPort() (int32, error) {
+	if dp.config.Kubelet.Port != 0 {
+		dp.logger.Debugf("Setting Port %d as specified by user config", dp.config.Kubelet.Port)
+		return dp.config.Kubelet.Port, nil
+	}
+
+	//We pay the price of a single call getting a node to avoid asking the user the Kubelet port if different from the standard one
+	node, err := dp.kc.CoreV1().Nodes().Get(context.Background(), dp.config.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return 0, fmt.Errorf("getting node %q: %w", dp.config.NodeName, err)
+	}
+
+	port := node.Status.DaemonEndpoints.KubeletEndpoint.Port
```

Comment:
Just to be sure, none of these fields is a pointer, right?

Reply:
no pointers